syntax = "proto2";

package market;

// Dispersy messages
message UserRequestMessage {
    required User user = 1;
}

message UserResponseMessage {
    required User user = 1;
}

message LoanRequestMessage {
    required LoanRequest loan_request = 1;
    required BorrowersProfile borrowers_profile = 2;
}

message LoanRejectMessage {
    required uint32 loan_request_id = 1;
    required bytes loan_request_user_id = 2;
}

message MortgageOfferMessage {
    required uint32 loan_request_id = 1;
    required bytes loan_request_user_id = 2;
    required Mortgage mortgage = 3;
}

message MortgageAcceptMessage {
    required uint32 mortgage_id = 1;
    required bytes mortgage_user_id = 2;
}

message MortgageRejectMessage {
    required uint32 mortgage_id = 1;
    required bytes mortgage_user_id = 2;
}

message InvestmentOfferMessage {
    required Investment investment = 1;
    required Profile investor_profile = 2;
}

message InvestmentAcceptMessage {
    required uint32 investment_id = 1;
    required bytes investment_user_id = 2;
    required BorrowersProfile borrowers_profile = 3;
}

message InvestmentRejectMessage {
    required uint32 investment_id = 1;
    required bytes investment_user_id = 2;
}

message CampaignUpdateMessage {
    required Campaign campaign = 1;
    required Mortgage mortgage = 2;
    optional Investment investment = 3;
}

message SignatureRequest {
    required uint32 identifier = 1;
    required Block block = 2;
}

message SignatureResponse {
    required uint32 identifier = 1;
    required Block block = 2;
}

// Objects that are included in the Dispersy messages
message LoanRequest {
    required uint32 id = 1;
    required bytes user_id = 2;
    required House house = 3;
    required uint32 mortgage_type = 4;
    required bytes bank_id = 5;
    required string description = 6;
    required double amount_wanted = 7;
    required uint32 status = 8;
}

message House {
    required string postal_code = 1;
    required string house_number = 2;
    required string address = 3;
    required double price = 4;
    required string url = 5;
    required string seller_phone_number = 6;
    required string seller_email = 7;
}

message Profile {
    required string first_name = 1;
    required string last_name = 2;
    required string email = 3;
    required string iban = 4;
    required string phone_number = 5;
}

message BorrowersProfile {
    required string first_name = 1;
    required string last_name = 2;
    required string email = 3;
    required string iban = 4;
    required string phone_number = 5;
    required string current_postal_code = 6;
    required string current_house_number = 7;
    required string current_address = 8;
    repeated string document_list = 9;
}

message Investment {
    required uint32 id = 1;
    required bytes user_id = 2;
    required double amount = 3;
    required uint32 duration = 4;
    required double interest_rate = 5;
    required uint32 campaign_id = 6;
    required bytes campaign_user_id = 7;
    required uint32 status = 8;
}

message Mortgage {
    required uint32 id = 1;
    required bytes user_id = 2;
    required bytes bank_id = 3;
    required House house = 4;
    required double amount = 5;
    required double bank_amount = 6;
    required uint32 mortgage_type = 7;
    required double interest_rate = 8;
    required double max_invest_rate = 9;
    required double default_rate = 10;
    required uint32 duration = 11;
    required string risk = 12;
    required uint32 status = 13;
    optional Campaign campaign = 14;
}

message Campaign {
    required uint32 id = 1;
    required bytes user_id = 2;
    required uint32 mortgage_id = 3;
    required bytes mortgage_user_id = 4;
    required double amount = 5;
    required int64 end_time = 6;
    required bool completed = 7;
    repeated Investment investments = 8;
}

message User {
    required bytes id = 1;
    required uint32 role = 2;
    optional Profile profile = 3;
}

message Block {
    required bytes signee = 1;
    required bytes document = 2;
    required int32 sequence_number_signee = 3;
    required bytes previous_hash_signee = 4;
    required bytes signature = 5;
}
